python/src/worker/tasks/marker_heatmap.py

```diff
@@ -42,52 +43,59 @@ def _format_request(self):
     def compute(self):
         request = self._format_request()
 
-        r = requests.post(
+        response = requests.post(
             f"{config.R_WORKER_URL}/v0/runMarkerHeatmap",
             headers={"content-type": "application/json"},
             data=json.dumps(request),
         )
-        # raise an exception if an HTTPError if one occurred because otherwise r.json() will fail
-        r.raise_for_status()
-        resultR = r.json()
-        truncatedR = resultR["truncatedExpression"]
-        resultR = resultR["rawExpression"]
+
+        # raise an exception if an HTTPError occurred
+        # as otherwise response.json() will fail
+        response.raise_for_status()
+        json_response = response.json()
+
+        error = json_response.get("error", False)
+        if error:
+            user_message = error.get("user_message", "")
+            err_code = error.get("error_code", "")
+            raise WorkerException(err_code, user_message)
+
+        response_data = json_response.get("data")
+
+        truncatedExpression = response_data["truncatedExpression"]
+        rawExpression = response_data["rawExpression"]
         result = {}
         data = {}
         order = []
-        if not len(resultR):
-            result[genes[0]] = {
-                "error": 404,
-                "message": "Gene {} not found!".format(genes[0]),
+
+        for gene in rawExpression.keys():
+            view = rawExpression[gene]
```

Comment:
this code is duplicated in the other heatmap move it into a function and call it from both places

Reply:
refactored